src/librustc_trait_selection/traits/coherence.rs

```diff
@@ -378,19 +379,25 @@ fn orphan_check_trait_ref<'tcx>(
         ty: Ty<'tcx>,
         in_crate: InCrate,
     ) -> Vec<Ty<'tcx>> {
-        if fundamental_ty(ty) && ty_is_non_local(ty, in_crate).is_some() {
-            ty.walk_shallow().flat_map(|ty| uncover_fundamental_ty(tcx, ty, in_crate)).collect()
-        } else {
-            vec![ty]
+        if let Some(inner_tys) = fundamental_ty_inner_tys(tcx, ty) {
+            // FIXME(eddyb) figure out if this is redundant with `ty_is_non_local`,
+            // or maybe if this should be calling `ty_is_non_local_constructor`.
+            if ty_is_non_local(tcx, ty, in_crate).is_some() {
```

Comment:
Nit: can we re-order the `if` statements? It feels surprising to me to introduce `inner_tys` and then only conditionally use it